packages/visibility_detector/lib/src/render_visibility_detector.dart

```diff
@@ -34,33 +98,220 @@ class RenderVisibilityDetector extends RenderProxyBox {
+        clip = clip.intersect(MatrixUtils.transformRect(transform, parentClip));
+      }
+      parent.applyPaintTransform(child, transform);
+    }
+    return VisibilityInfo.fromRects(
+      key: key,
+      widgetBounds: MatrixUtils.transformRect(transform, bounds),
+      clipRect: clip,
+    );
+  }
+
+  /// Used to get the bounds of the render object when it is time to update
+  /// clients about visibility.
+  Rect get bounds;
 
-  /// See [RenderObject.paint].
   @override
   void paint(PaintingContext context, Offset offset) {
-    if (onVisibilityChanged == null) {
-      // No need to create a [VisibilityDetectorLayer].  However, in case one
-      // already exists, remove all cached data for it so that we won't fire
-      // visibility callbacks when the layer is removed.
-      VisibilityDetectorLayer.forget(key);
-      super.paint(context, offset);
-      return;
+    if (onVisibilityChanged != null) {
+      _compositionCallbackCanceller?.call();
+      _compositionCallbackCanceller =
+          context.addCompositionCallback((Layer layer) {
+        assert(!debugDisposed!);
+        final ContainerLayer? container =
+            layer is ContainerLayer ? layer : layer.parent;
+        _scheduleUpdate(container);
+      });
     }
+    super.paint(context, offset);
+  }
 
-    final layer = VisibilityDetectorLayer(
-        key: key,
-        widgetOffset: Offset.zero,
-        widgetSize: semanticBounds.size,
-        paintOffset: offset,
-        onVisibilityChanged: onVisibilityChanged!);
-    context.pushLayer(layer, super.paint, offset);
+  bool _disposed = false;
+  @override
+  void dispose() {
+    _compositionCallbackCanceller?.call();
+    _compositionCallbackCanceller = null;
+    _disposed = true;
+    super.dispose();
+  }
+}
+
+/// The [RenderObject] corresponding to the [VisibilityDetector] widget.
+class RenderVisibilityDetector extends RenderProxyBox
+    with RenderVisibilityDetectorBase {
+  /// Constructor.  See the corresponding properties for parameter details.
+  RenderVisibilityDetector({
+    RenderBox? child,
+    required this.key,
+    required VisibilityChangedCallback? onVisibilityChanged,
+  }) : assert(key != null),
+        _onVisibilityChanged = onVisibilityChanged,
+        super(child);
+
+  @override
+  final Key key;
+
+  @override
+  Rect get bounds => semanticBounds;
+}
+
+/// The [RenderObject] corresponding to the [SliverVisibilityDetector] widget.
+///
+/// [RenderSliverVisibilityDetector] is a bridge between
+/// [SliverVisibilityDetector] and [VisibilityDetectorLayer].
+class RenderSliverVisibilityDetector extends RenderProxySliver
+    with RenderVisibilityDetectorBase {
+  /// Constructor.  See the corresponding properties for parameter details.
+  RenderSliverVisibilityDetector({
+    RenderSliver? sliver,
+    required this.key,
+    required VisibilityChangedCallback? onVisibilityChanged,
+  }) : super(sliver) {
+    _onVisibilityChanged = onVisibilityChanged;
+  }
```

Comment:
similar to above:

```suggestion
  }) : assert(key != null),
        _onVisibilityChanged = onVisibilityChanged,
        super(child);
```

Reply:
Doesn't work - because the private field is on a mixin. Maybe this is a dart bug?

Reply:
I see. Interesting.
Maybe add the assert, though just for symmetry with the other constructor above..